internal/app/server/server.go

```diff
@@ -66,55 +102,63 @@ func Run(ctx context.Context, args []string, in io.ReadCloser, out, err io.Write
 	}).Handler(mux)
 
 	// Create servers
-	h1Server := newH1Server(*h1Port, corsHandler)
-	h2Server := newH2Server(*h2Port, mux)
-	done := make(chan os.Signal, 1)
-	errs := make(chan error, 2)
-	signal.Notify(done, os.Interrupt, syscall.SIGINT, syscall.SIGTERM)
-	go func() {
-		err := h1Server.ListenAndServe()
-		if err != nil && !errors.Is(err, http.ErrServerClosed) {
-			errs <- err
-		}
-	}()
+	var server *http.Server
+	if req.HttpVersion == v1alpha1.HTTPVersion_HTTP_VERSION_1 {
+		server = newH1Server(corsHandler)
+	} else if req.HttpVersion == v1alpha1.HTTPVersion_HTTP_VERSION_2 {
+		server = newH2Server(mux)
+	} else if req.HttpVersion == v1alpha1.HTTPVersion_HTTP_VERSION_3 {
+		return nil, errors.New("HTTP/3 is not yet supported")
+	} else {
+		return nil, errors.New("an HTTP version must be specifed.")
+	}
+	ln, err := net.Listen("tcp", net.JoinHostPort(defaultHost, defaultPort))
+	if err != nil {
+		return nil, err
+	}
+	resp := &v1alpha1.ServerCompatResponse{}
+	errs := make(chan error, 1)
 	go func() {
-		err := h2Server.ListenAndServe()
+		err := server.Serve(ln)
 		if err != nil && !errors.Is(err, http.ErrServerClosed) {
 			errs <- err
 		}
 	}()
 
-	fmt.Printf("HTTP/1.1 server listening on port %s\nHTTP/2 server listening on port %s", *h1Port, *h2Port)
-
+	// TODO - Is there a race condition here where we reach the default case before any error is sent
+	// on the errs channel?
```

Comment:
It doesn't look quite right to me.

I'm not sure if it's idiomatic, but you can send `nil` on the channel on success and an error on failure. You wouldn't need select.

Reply:
The problem is that `server.Serve` blocks, so you can't send anything after. It's almost like I need a small delay before attempting to read from the `errs` channel to give the server time to start up and verify no errors occurred.

Reply:
Errors at this stage are very rare. And, as you point out, there is nothing that can be done here since `Serve` blocks until the server shuts down if it is successful.

As currently implemented, the `default` case will almost certainly be taken: even if `Serve` returns an error, it is unlikely for the goroutine to have started, error returned, and error pushed into channel before we get to this point. So, since the `errs` channel is not ready, it will fall to the default case.

But another big problem here is that this immediately returns, which causes `Run` to return, which then cases `main` to return, which means that the server immediately stops and the process exits.

A better approach is to have this function return `(*http.Server, error)`. Then the caller can write the response to stdout on success and _then_ call `server.Serve`, but not in a goroutine, but blocking. That way, if `Serve` does unexpectedly return an error, it will cause `Run` to return an error, which should cause the program to exit with an error code, which I think is what would be expected.